### PR TITLE
Lay groundwork for label-free backports

### LINKED ISSUE #19: Add "/trop run backport-to $branchname"
For when you have a one-off need and don't want to pollute the label namespace.

Inspired by a joking suggestion that I made to @MarshallOfSound last night that we fix
`electron/set-appropriate-defaults-for-webview-options` and
`electron/set-appropriate-defaults-for-webview-options_1-8-x`
with a meta-PR that had the label 
`target/set-appropriate-defaults-for-webview-options_1-8-x`

### TODO

- [x] get feedback from @codebytere and @MarshallOfSound
- [x] move user authentication logic to outside of `/trop run backport` command handler
- [x] add code to recognize `/trop run backport-to branch $branchname` to pull out the branchname
- [x] change backport/utils:
   - [x] two forms of backportPR, one taking a label and the other a branchname
   - [x] if no label is provided, don't try to remove it
   - [x] if a branch is provided, don't try to guess the branchname from the label
- [x] change src/index.js:
   - [x] add a sibling func next to `backportAllLabels()` called `backportTo(branchname)` that prepares `context` and passes the work along to `utils`
   - [ ] finish the work in `robot.on('issue_comment.created')` to call `backportTo()`
- [x] tests

### PR DESCRIPTION
**Pull request by ckerr**
Lays the groundwork for #19, but does not actually plug in the new `backport-to` command yet.

I'm stopping here for now because (a) @codebytere also has some logic to refactor and I want to prevent dueling refactors, and (b) I really ought to be working on the gtk theming issue anyway :+1:

### CODE REVIEW

File: src/backport/runner.ts [hunk 1]
```diff
@@ -56,7 +56,7 @@ const setUpRemotes = async (details: { slug: string, remotes: { name: string, va
   return { success: true };
 }
 
-const backportPR = async (details: { slug: string, targetRemote: string, targetBranch: string, tempRemote: string, tempBranch: string, commits: string[] }) => {
+const backportToLabel = async (details: { slug: string, targetRemote: string, targetBranch: string, tempRemote: string, tempBranch: string, commits: string[] }) => {
```
Comment (MarshallOfSound) [inline src/backport/runner.ts]: I know I wrote this bit 😆 But reading through want this method actually does I would in hindsight call it `backportCommitsToBranch` 🤔

File: src/backport/utils.ts [hunk 2]
```diff
@@ -44,16 +44,17 @@ const tellRunnerTo = async (what: string, payload: any) => {
...
-export const backportPR = async (robot: Probot, context: ProbotContext<PullRequestEvent>, label: Label) => {
-  const config = await context.config('config.yml');
-  const targetLabelPrefix = config.targetLabelPrefix || TARGET_LABEL_PREFIX;
-  const mergedLabelPrefix = config.mergedLabelPrefix || MERGED_LABEL_PREFIX;
+const backportImpl = async (robot: Probot,
+                            context: ProbotContext<PullRequestEvent>,
+                            targetBranch: string,
+                            labelToRemove?: string,
+                            labelToAdd?: string) => {
+  if (!targetBranch)
```
Comment (MarshallOfSound) [inline src/backport/utils.ts]: Thanks to the gloriousness of typescript we don't need this null check, `targetBranch` must always be a valid string

File: src/backport/utils.ts [hunk 3]
```diff
@@ -217,3 +224,29 @@ export const backportPR = async (robot: Probot, context: ProbotContext<PullReque
     }) as any);
   });
 }
+
+const getLabelPrefixes = async (context: ProbotContext<PullRequestEvent>) => {
```
Comment (MarshallOfSound) [inline src/backport/utils.ts]: `context` here doesn't have to be a `PullRequestEvent` context.  The type here could be something like `ProbotContext<any>` to allow any context 👍

File: src/backport/utils.ts [hunk 4]
```diff
@@ -217,3 +224,29 @@ export const backportPR = async (robot: Probot, context: ProbotContext<PullReque
...
+  const config = await context.config('config.yml');
+  const target = config.targetLabelPrefix || TARGET_LABEL_PREFIX;
+  const merged = config.mergedLabelPrefix || MERGED_LABEL_PREFIX;
+  return { target, merged }
+}
+
+export const backportToLabel = async (robot: Probot, context: ProbotContext<PullRequestEvent>, label: Label) => {
+  const labelPrefixes = await getLabelPrefixes(context);
+  if (!label.name.startsWith(labelPrefixes.target))
+    throw new Error(`Label '${label.name}' does not begin with '${labelPrefixes.target}'`);
```
Comment (MarshallOfSound) [inline src/backport/utils.ts]: This shouldn't throw an error, it should just silently fail IMO.  PR's can be tagged as things like `feature` or `osr` and throwing an error will break things here, we really just need to skip labels that don't match that pattern 👍

File: src/backport/utils.ts [hunk 5]
```diff
@@ -44,16 +44,17 @@ const tellRunnerTo = async (what: string, payload: any) => {
...
-  const config = await context.config('config.yml');
-  const targetLabelPrefix = config.targetLabelPrefix || TARGET_LABEL_PREFIX;
-  const mergedLabelPrefix = config.mergedLabelPrefix || MERGED_LABEL_PREFIX;
+const backportImpl = async (robot: Probot,
+                            context: ProbotContext<PullRequestEvent>,
+                            targetBranch: string,
+                            labelToRemove?: string,
+                            labelToAdd?: string) => {
+  if (!targetBranch)
+    throw new Error('Nothing to do');
```
Comment (MarshallOfSound) [inline src/backport/utils.ts]: Also yeah, let's avoid throwing errors, they break the promise chain.  If we actually just want to stop doing something just log out and `return`

File: src/backport/utils.ts [hunk 6]
```diff
@@ -187,21 +185,27 @@ export const backportPR = async (robot: Probot, context: ProbotContext<PullReque
       maintainer_can_modify: false,
     }))).data;
 
-    log('Adding handy comment and updating labels')
+    log('Adding breadcrumb comment')
```
Comment (codebytere) [inline src/backport/utils.ts]: why do we `log` in some places and `robot.log` in others?

File: src/index.js [hunk 7]
```diff
@@ -144,31 +144,45 @@ module.exports = async (robot) => {
...
+
+      if (cmd === backportCmd) {
+        await context.github.issues.createComment(context.repo({
+          number: payload.issue.number,
+          body: `The backport process for this PR has been manually initiated, here we go! :D`
+        }))
+        backportAllLabels(context, pr)
+      } else if (cmd.startsWith(backportToCmd)) {
+        const targetBranch = cmd.slice(backportToCmd.length)
+        robot.log('backport-to ' + targetBranch)
```
Comment (codebytere) [inline src/index.js]: (nit) template literals are preferred

File: src/index.js [hunk 8]
```diff
@@ -144,31 +144,45 @@ module.exports = async (robot) => {
...
+        // TODO: sanitize this input - does the branch exist?
```
Comment (codebytere) [inline src/index.js]: I believe for this you'd hit https://developer.github.com/v3/git/refs/#get-a-reference and ensure that a single object was returned and that it has a `ref` key

File: src/backport/utils.ts (hunk 6 above)
Comment (ckerr) [reply src/backport/utils.ts]: `backportImpl()` has this nice wrapper that adds `username/repo` to the front of each log line, but it's not shared everywhere. The quick & dirty alternative would be to copypaste that wrapper here, but I didn't really like the code duplication so I left it alone.

IMO the Right Thing would be a log wrapper that we can dependency-inject and pass around, but  that's better left to another PR

**Comment by codebytere** [comment]
@MarshallOfSound do you have any changes you'd want to make to this? i'm comfortable merging this as-is and then opening a separate PR for the actual plugin

**Pull request merged**